server/service/http/message/query/query_structure.rs

```diff
@@ -55,67 +55,125 @@ impl<'a, Snapshot: ReadableSnapshot> QueryStructureContext<'a, Snapshot> {
+        role: QueryStructureVertexResponse,
+        exactness: QueryStructureConstraintExactness,
+    },
+    Plays {
+        player: QueryStructureVertexResponse,
+        role: QueryStructureVertexResponse,
+        exactness: QueryStructureConstraintExactness,
+    },
 
-    IsaExact,
-    SubExact,
+    FunctionCall {
+        name: String,
+        assigned: Vec<QueryStructureVertexResponse>,
+        arguments: Vec<QueryStructureVertexResponse>,
+    },
+    Expression {
+        text: String,
+        assigned: Vec<QueryStructureVertexResponse>,
+        arguments: Vec<QueryStructureVertexResponse>,
+    },
+}
 
-    Assigned(String),
-    Argument(String),
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct QueryConstraintIsaBase {
+    instance: QueryStructureVertexResponse,
+    r#type: QueryStructureVertexResponse,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct QueryConstraintSubBase {
+    subtype: QueryStructureVertexResponse,
+    supertype: QueryStructureVertexResponse,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub enum QueryStructureConstraintExactness {
+    Exact,
+    Subtypes,
 }
 
 #[derive(Debug, Clone, Copy, Serialize, Deserialize)]
 #[serde(rename_all = "camelCase")]
-struct QueryStructureEdgeSpan {
+struct QueryStructureConstraintSpan {
     begin: usize,
     end: usize,
 }
 
 #[derive(Debug, Serialize, Deserialize)]
 #[serde(rename_all = "camelCase")]
-pub struct QueryStructureEdgeResponse {
-    r#type: QueryStructureEdgeTypeResponse,
-    from: QueryStructureVertexResponse,
-    to: QueryStructureVertexResponse,
-    span: Option<QueryStructureEdgeSpan>,
+pub struct QueryStructureConstraintResponse {
+    text_span: Option<QueryStructureConstraintSpan>,
+    #[serde(flatten)]
+    constraint: QueryStructureConstraint,
 }
 
 #[derive(Debug, Serialize, Deserialize, Clone)]
-#[serde(rename_all = "camelCase", tag = "kind", content = "value")]
+#[serde(rename_all = "camelCase", tag = "tag")]
 pub enum QueryStructureVertexResponse {
     Variable { variable: String },
-    Label(serde_json::Value),
+    Label { r#type: serde_json::Value },
     Value(ValueResponse),
     Expression { repr: String },
     FunctionCall { repr: String },
     UnavailableVariable { variable: String },
 }
 
+#[derive(Debug, Serialize, Deserialize, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct QueryStructureRolePlayerResponse {
+    player: QueryStructureVertexResponse,
+    role: QueryStructureVertexResponse,
+    span: Option<QueryStructureConstraintSpan>,
+}
+
 pub(crate) fn encode_query_structure(
     snapshot: &impl ReadableSnapshot,
     type_manager: &TypeManager,
     query_structure: &QueryStructure,
 ) -> Result<QueryStructureResponse, Box<ConceptReadError>> {
-    let branches = query_structure
+    let blocks = query_structure
         .parametrised_structure
         .branches
```

Comment:
I can propagate this all the way if needed, but maybe after the first round of review.

Reply:
Yes let's do that in the next round